File: CCC/ccc14s3.java
import java.util.Scanner;
import java.util.Stack;

public class ccc14s3 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        boolean isPossible = true;
        int test = input.nextInt();
        for(int i = 0; i < test; i++){
        int counter;
            counter = 1;
            int numCars = input.nextInt();
            Stack<Integer> mountain = new Stack<Integer>();
            Stack<Integer> branch = new Stack<Integer>();
            for(int j = 1; j <= numCars; j++){
                mountain.push(input.nextInt());
            }
            
            while(!mountain.isEmpty()){
                if(mountain.peek()==counter){
                    mountain.pop();
                    counter++;
                   } 
                else if(!branch.isEmpty() && branch.peek()==counter){
                    branch.pop();
                    counter++;
                }else{
                    branch.push(mountain.pop());
                }
            }
                isPossible = true;
                while(!branch.isEmpty()){
                    if(branch.peek()==counter){
                        branch.pop();
                        counter++;
                    }else{
                        isPossible = false;
                        branch.pop();
                    }
                } 
                if(isPossible){
                    System.out.println("Y");
                }else{
                    System.out.println("N");
                }
    }
}
}

File: CCC/ccc14s3.cpp
#include <bits/stdc++.h>
using namespace std;
int T,N,M;
int main(){
  scanf("%d",&T);
  for (int i=0;i<T;i++){
    int counter=1;
    stack <int> mountain;
    stack <int> branch; 
    scanf("%d",&N);
    for (int i=0;i<N;i++){
      scanf("%d",&M);
      mountain.push(M);
    }
    while (!mountain.empty()){
      if (mountain.top()==counter){
        mountain.pop();
        counter++;
      }
      else if (!branch.empty()&&branch.top()==counter){
        branch.pop();
        counter++;
      }
      else{
        branch.push(mountain.top());
        mountain.pop();
      }
    }
    bool memed=false;
    while (!branch.empty()){
      if (branch.top()==counter){
        branch.pop();
        counter++;
      }
      else{
        memed=true;
        break;
      }
    }
    printf(memed? "N\n" : "Y\n");
  }
}